Clean-up and unit test the renaming level classes

romainbrenguier opened the pull request:
This removes unnecessary inheritance, and makes the public interface of symex_level1t smaller so that it can be more significantly unit-tested. In particular we ensure that it is only used to map identifiers which come from L0 renamed ssa_exprt.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -796,28 +796,18 @@ ssa_exprt goto_symex_statet::add_object(
-    }
-
-    level1.current_names.replace(l0_name, std::make_pair(ssa, l1_index));
+    if(!frame.old_level1.has(renamed))
```
Normally I'd expect an `insert` in C++ to be a no-op if the key is not present

romainbrenguier replied:
You mean if it's present?
Instead `sharing_mapt` has a precondition what ensures we never insert twice the same key, it seems more secure to abort the program there rather than silently doing nothing. The alternative solution would be to give another name to insert.

smowton replied:
Ah crap forgot this was `sharing_map`. Indeed it surely ought to have the same semantics as `std::map::insert` and its current `insert` should be called something like `insert_fresh`

romainbrenguier replied:
It makes sense but I would leave that for another pull request since this is related more to sharing map than renamings

smowton commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -97,6 +136,12 @@ void symex_level1t::restore_from(const current_namest &other)
   }
 }
 
+unsigned symex_level2t::current_count(const irep_idt &identifier) const
```
This name sounds too much like `std::map::count`, how about `latest_renaming_index`?

romainbrenguier replied:
Maybe `latest_index`? "renaming" is kind of implied when talking about `symex_level2t`

smowton commented on src/goto-symex/symex_start_thread.cpp:
```diff
@@ -67,16 +67,16 @@ void goto_symext::symex_start_thread(statet &state)
-    lhs.set_level_0(t);
-    const irep_idt &l0_name = lhs.get_identifier();
+    const renamedt<ssa_exprt, L0> l0_lhs =
+      symex_level0t{}(std::move(lhs), ns , t);
```
So it's default constructed then `operator()`'d? Could that be improved? Perhaps comment what's going on?

romainbrenguier replied:
I've made this functor a function instead so it should be better

smowton reviewed (approved):
Mostly lgtm just a few minor comments

romainbrenguier committed:
Move renamedt class to its own file

Make navigating the classes easier.

romainbrenguier committed:
Apply clang-format suggestion to renamed.h

This is kept separated from the commit moving the class so that the diff
of the move is cleaner.

romainbrenguier committed:
Deprecate unused function

This is not tested, so we cannot guarantee it works.

romainbrenguier committed:
Replace symex_renaming_levelt inheritance by field

The inheritance is not actually used as only the destructor is virtual.
The code is simpler without inheritance.

romainbrenguier committed:
Document symex_level0t

Documentation of the behaviour of the operator was missing.

romainbrenguier committed:
Document symex_level1t

The documentation for the operator() was missing.

romainbrenguier committed:
Document symex_level2t

The documentation of the operator() was missing.

romainbrenguier committed:
Add insert and insert_or_replace level1t methods

This makes the interaction with the map of l1 indexes easier.

romainbrenguier committed:
Make current_names private in symex_level1t

This ensures the map is updated in a consistent way, in particular the
identifier used as keys should always belong to the same renaming level.

romainbrenguier committed:
Use symex_level0t::operator() instead of set_level_0

Calling the renaming makes additional checks compared to the low level
update of the ssa_exprt and the return type is more informative.

romainbrenguier committed:
Make insert_(or_replace) take renamedt argument

This ensures that the identifier we insert in the map correspond to
ssa_exprt that have been renamed to level 0.

romainbrenguier committed:
Add unit test for symex_level0t

This tests the behavior on different SSA expressions.

romainbrenguier committed:
Add unit test for symex_level1t

This tests the behaviour of insert, insert_or_replace and the
operator().

romainbrenguier committed:
Add a has method to symex_level1t

This will be used when we make the map for old L1 name be a
symex_level1t as well.

romainbrenguier committed:
Make restore_from take symex_level1t

This ensures we don't accidentaly mix maps from different levels.

romainbrenguier committed:
Unit test the symex_level1t::has method

Add condition in the unit test to check the result is as expected.

romainbrenguier committed:
Rename current_count method to latest_index

current_count could sound related to count which has a different meaning
from what this method returns.

romainbrenguier committed:
Make symex_level0 a function

A functor without state is just a function.

owen-mc-diffblue commented on src/goto-symex/renaming_level.h:
```diff
@@ -56,7 +56,14 @@ struct symex_level0t
 /// This is to preserve locality in case of recursion
 struct symex_level1t
 {
-  symex_renaming_levelt current_names;
```
This deletion should be in the next commit (" Make current_names private in symex_level1t"), I think

owen-mc-diffblue commented on unit/goto-symex/symex_level1.cpp:
```diff
@@ -0,0 +1,68 @@
+    {
+      symex_level1.insert(l0_symbol, 12134);
+      auto renamed = symex_level1(l0_symbol);
+      REQUIRE(renamed.get().get_identifier() == "foo!50@12134");
```
I'd expect some `THEN` sections

owen-mc-diffblue commented on unit/goto-symex/symex_level1.cpp:
```diff
@@ -0,0 +1,68 @@
+    WHEN("insert_or_replace is called with a symbol not already inserted")
+    {
+      auto old = symex_level1.insert_or_replace(l0_symbol, 20051);
+      ;
```
Stray semicolon

peterschrammel commented on src/goto-symex/renaming_level.h:
```diff
@@ -42,6 +42,9 @@ void get_variables(
 {
+  /// Rename \p ssa_expr using \p thread_nr as L0 tag, unless \p ssa_expr is
+  /// a guard, a shared variable or a function. \p ns is queried to decide
+  /// whether we are in one of this cases.
```
typo `these`

peterschrammel commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -798,23 +798,13 @@ ssa_exprt goto_symex_statet::add_object(
   ssa_exprt ssa = rename_ssa<L0>(ssa_exprt{expr}, ns).get();
   const irep_idt l0_name = ssa.get_identifier();
-  const std::size_t l1_index = index_generator(l0_name);
+  const auto l1_index = narrow_cast<unsigned>(index_generator(l0_name));
```
Why `unsigned` and not `size_t`?

romainbrenguier replied:
`symex_renaming_levelt` uses `unsigned` for indexes

peterschrammel commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -56,6 +56,28 @@ operator()(ssa_exprt ssa_expr, const namespacet &ns, unsigned thread_nr) const
 
+void symex_level1t::insert(ssa_exprt ssa, unsigned index)
+{
+  const irep_idt identifier = ssa.get_identifier();
```
`const &` ? (several occurrences)

romainbrenguier replied:
the problem is `ssa` is moved below, and the move will happen before we use the identifier, so we need a copy

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9992388747163a707bfd6a5a3afd501eed8ef871).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114705446

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 530c35b8e45c72217225c3fc7b78747355c8698f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114711776
